getDisplayMedia does not require a list of every screen and window

eric-carlson opened the pull request:
#### 1ad6bf1f3f6c8122574d79e9ae8bb53a2a61669a
<pre>
getDisplayMedia does not require a list of every screen and window
<a href="https://bugs.webkit.org/show_bug.cgi?id=251719">https://bugs.webkit.org/show_bug.cgi?id=251719</a>
rdar://105018387

Reviewed by Youenn Fablet.

getDisplayMedia always requires a prompt and an explicit user selection, so we will
never need to build a list of every window and screen.

* Source/WebCore/platform/mediastream/DisplayCaptureManager.h:
(WebCore::DisplayCaptureManager::windowDevices): Deleted.
* Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp:
(WebCore::RealtimeMediaSource::setIntrinsicSize):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCenter.cpp:
(WebCore::RealtimeMediaSourceCenter::getMediaStreamDevices):
(WebCore::RealtimeMediaSourceCenter::enumerateDevices):
(WebCore::RealtimeMediaSourceCenter::validateRequestConstraints):
(WebCore::RealtimeMediaSourceCenter::validateRequestConstraintsAfterEnumeration):
(WebCore::RealtimeMediaSourceCenter::getDisplayMediaDevices): Deleted.
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCenter.h:
* Source/WebCore/platform/mediastream/cocoa/DisplayCaptureSourceCocoa.cpp:
(WebCore::DisplayCaptureSourceCocoa::updateFrameSize):
(WebCore::DisplayCaptureSourceCocoa::capturerConfigurationChanged):
* Source/WebCore/platform/mediastream/cocoa/DisplayCaptureSourceCocoa.h:
* Source/WebCore/platform/mediastream/mac/DisplayCaptureManagerCocoa.cpp:
(WebCore::DisplayCaptureManagerCocoa::captureDevices):
(WebCore::DisplayCaptureManagerCocoa::updateDisplayCaptureDevices): Deleted.
(WebCore::DisplayCaptureManagerCocoa::updateWindowCaptureDevices): Deleted.
(WebCore::DisplayCaptureManagerCocoa::windowDevices): Deleted.
* Source/WebCore/platform/mediastream/mac/DisplayCaptureManagerCocoa.h:
* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.mm:
(WebCore::ScreenCaptureKitCaptureSource::startContentStream):
(WebCore::ScreenCaptureKitCaptureSource::intrinsicSize const):
(WebCore::ScreenCaptureKitCaptureSource::streamDidOutputSampleBuffer):
(WebCore::ScreenCaptureKitCaptureSource::windowCaptureDeviceWithPersistentID):
(WebCore::ScreenCaptureKitCaptureSource::captureDeviceWithPersistentID): Deleted.
(WebCore::ScreenCaptureKitCaptureSource::screenCaptureDevices): Deleted.
(WebCore::ScreenCaptureKitCaptureSource::windowCaptureDevices): Deleted.
(WebCore::ScreenCaptureKitCaptureSource::windowDevices): Deleted.
(WebCore::forEachNSWindow): Deleted.
* Source/WebCore/platform/mock/MockRealtimeMediaSourceCenter.cpp:
(WebCore::MockRealtimeMediaSourceCenter::MockDisplayCaptureDeviceManager::windowDevices): Deleted.
* Source/WebCore/platform/mock/MockRealtimeMediaSourceCenter.h:
* Source/WebKit/UIProcess/API/C/WKUserMediaPermissionRequest.cpp:
(WKUserMediaPermissionRequestRequiresDisplayCapture):
* Source/WebKit/UIProcess/API/C/WKUserMediaPermissionRequest.h:
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
(WebKit::UserMediaPermissionRequestManagerProxy::finishGrantingRequest):
(WebKit::UserMediaPermissionRequestManagerProxy::getRequestAction):
(WebKit::UserMediaPermissionRequestManagerProxy::processUserMediaPermissionRequest):
(WebKit::UserMediaPermissionRequestManagerProxy::processUserMediaPermissionValidRequest):
* Source/WebKit/UIProcess/UserMediaPermissionRequestProxy.h:
(WebKit::UserMediaPermissionRequestProxy::requiresDisplayCapture const):
(WebKit::UserMediaPermissionRequestProxy::requiresDisplayCaptureWithAudio const):
* Source/WebKit/UIProcess/UserMediaProcessManager.cpp:
(WebKit::UserMediaProcessManager::willCreateMediaStream):
* Source/WebKit/UIProcess/UserMediaProcessManager.h:
* Source/WebKit/UIProcess/mac/DisplayCaptureSessionManager.mm:
(WebKit::DisplayCaptureSessionManager::showWindowPicker):
(WebKit::DisplayCaptureSessionManager::showScreenPicker):
(WebKit::alertForWindowSelection): Deleted.
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::decidePolicyForUserMediaPermissionRequestIfPossible):

Canonical link: <a href="https://commits.webkit.org/259969@main">https://commits.webkit.org/259969@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/68fdb74b48f3e50010fc74eca0680668d4385779

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106588 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/77/builds/15598 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/39391 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115772 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/115406 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/110494 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/17095 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6828 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98779 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112356 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/12989 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/39391 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/40549 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/94867 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/39391 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/82269 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/8824 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/39391 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9368 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/5991 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [❌ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14986 "Failed to checkout and rebase branch from PR 9625") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/39391 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6897 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/10904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/mediastream/RealtimeMediaSourceCenter.cpp:
```diff
@@ -197,25 +196,6 @@ void RealtimeMediaSourceCenter::triggerDevicesChangedObservers()
-void RealtimeMediaSourceCenter::getDisplayMediaDevices(const MediaStreamRequest& request, MediaDeviceHashSalts&& hashSalts, Vector<DeviceInfo>& displayDeviceInfo, String& firstInvalidConstraint)
-{
-    if (!request.videoConstraints.isValid)
-        return;
```
This is probably the part that we need to add in UIProcess to reject immediately when { video: false } or equivalent is given to getDisplayMedia.

eric-carlson replied:
This check was unnecessary, we reject in `MediaDevices::getDisplayMedia` if the constraints are invalid or over constrained.

youennf commented on Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.mm:
```diff
@@ -468,63 +470,11 @@ static void findSharableDevice(RetainPtr<SCShareableContent>&& shareableContent,
-
-    return { static_cast<int>(bounds.size.width), static_cast<int>(bounds.size.height) };
-#endif
+    return { 640, 480 };
```
Shouldn't we return 0,0?

Something like `if (!m_content) return { 0, 0 };`

eric-carlson replied:
Returning 0x0 regresses the fix for https://bugs.webkit.org/show_bug.cgi?id=247310. We should fix this, but lets do it in another patch.

youennf commented on Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
```diff
@@ -581,12 +578,18 @@ void UserMediaPermissionRequestManagerProxy::processUserMediaPermissionRequest()
                 return;
 
+            if (videoDevices.isEmpty() && request->requiresDisplayCapture() && m_page.preferences().mockCaptureDevicesEnabled())
+                videoDevices = WebCore::RealtimeMediaSourceCenter::singleton().displayCaptureFactory().displayCaptureDeviceManager().captureDevices();
```
Why do we need this? Can we move this code to mock capture code if needed?
Maybe in processUserMediaPermissionValidRequest, around lines 651/654?

eric-carlson replied:
We need this because when mock devices are active we don't necessarily go through the mock picker path. I will move it to `processUserMediaPermissionValidRequest`.

eric-carlson committed:
getDisplayMedia does not require a list of every screen and window
https://bugs.webkit.org/show_bug.cgi?id=251719
rdar://105018387

Reviewed by Youenn Fablet.

getDisplayMedia always requires a prompt and an explicit user selection, so we will
never need to build a list of every window and screen.

* Source/WebCore/platform/mediastream/DisplayCaptureManager.h:
(WebCore::DisplayCaptureManager::windowDevices): Deleted.
* Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp:
(WebCore::RealtimeMediaSource::setIntrinsicSize):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCenter.cpp:
(WebCore::RealtimeMediaSourceCenter::getMediaStreamDevices):
(WebCore::RealtimeMediaSourceCenter::enumerateDevices):
(WebCore::RealtimeMediaSourceCenter::validateRequestConstraints):
(WebCore::RealtimeMediaSourceCenter::validateRequestConstraintsAfterEnumeration):
(WebCore::RealtimeMediaSourceCenter::getDisplayMediaDevices): Deleted.
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCenter.h:
* Source/WebCore/platform/mediastream/cocoa/DisplayCaptureSourceCocoa.cpp:
(WebCore::DisplayCaptureSourceCocoa::updateFrameSize):
(WebCore::DisplayCaptureSourceCocoa::capturerConfigurationChanged):
* Source/WebCore/platform/mediastream/cocoa/DisplayCaptureSourceCocoa.h:
* Source/WebCore/platform/mediastream/mac/DisplayCaptureManagerCocoa.cpp:
(WebCore::DisplayCaptureManagerCocoa::captureDevices):
(WebCore::DisplayCaptureManagerCocoa::updateDisplayCaptureDevices): Deleted.
(WebCore::DisplayCaptureManagerCocoa::updateWindowCaptureDevices): Deleted.
(WebCore::DisplayCaptureManagerCocoa::windowDevices): Deleted.
* Source/WebCore/platform/mediastream/mac/DisplayCaptureManagerCocoa.h:
* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.mm:
(WebCore::ScreenCaptureKitCaptureSource::startContentStream):
(WebCore::ScreenCaptureKitCaptureSource::intrinsicSize const):
(WebCore::ScreenCaptureKitCaptureSource::streamDidOutputSampleBuffer):
(WebCore::ScreenCaptureKitCaptureSource::windowCaptureDeviceWithPersistentID):
(WebCore::ScreenCaptureKitCaptureSource::captureDeviceWithPersistentID): Deleted.
(WebCore::ScreenCaptureKitCaptureSource::screenCaptureDevices): Deleted.
(WebCore::ScreenCaptureKitCaptureSource::windowCaptureDevices): Deleted.
(WebCore::ScreenCaptureKitCaptureSource::windowDevices): Deleted.
(WebCore::forEachNSWindow): Deleted.
* Source/WebCore/platform/mock/MockRealtimeMediaSourceCenter.cpp:
(WebCore::MockRealtimeMediaSourceCenter::MockDisplayCaptureDeviceManager::windowDevices): Deleted.
* Source/WebCore/platform/mock/MockRealtimeMediaSourceCenter.h:
* Source/WebKit/UIProcess/API/C/WKUserMediaPermissionRequest.cpp:
(WKUserMediaPermissionRequestRequiresDisplayCapture):
* Source/WebKit/UIProcess/API/C/WKUserMediaPermissionRequest.h:
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
(WebKit::UserMediaPermissionRequestManagerProxy::finishGrantingRequest):
(WebKit::UserMediaPermissionRequestManagerProxy::getRequestAction):
(WebKit::UserMediaPermissionRequestManagerProxy::processUserMediaPermissionRequest):
(WebKit::UserMediaPermissionRequestManagerProxy::processUserMediaPermissionValidRequest):
* Source/WebKit/UIProcess/UserMediaPermissionRequestProxy.h:
(WebKit::UserMediaPermissionRequestProxy::requiresDisplayCapture const):
(WebKit::UserMediaPermissionRequestProxy::requiresDisplayCaptureWithAudio const):
* Source/WebKit/UIProcess/UserMediaProcessManager.cpp:
(WebKit::UserMediaProcessManager::willCreateMediaStream):
* Source/WebKit/UIProcess/UserMediaProcessManager.h:
* Source/WebKit/UIProcess/mac/DisplayCaptureSessionManager.mm:
(WebKit::DisplayCaptureSessionManager::showWindowPicker):
(WebKit::DisplayCaptureSessionManager::showScreenPicker):
(WebKit::alertForWindowSelection): Deleted.
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::decidePolicyForUserMediaPermissionRequestIfPossible):

Canonical link: https://commits.webkit.org/259969@main
